<launch>
    <!-- README: IF YOU WANT TO USE THIS LAUNCH FILE YOU SHOULD DO:
        git checkout simulador 
        INTO THE COLA2_TURBOT PACKAGE -->
    <!-- Params -->
    <!-- Params -->
    <arg name="place" default="andratx_exterior.yaml"/>
    <arg name="robot0" default="true"/>
    <arg name="robot1" default="true"/>
    <arg name="robot2" default="true"/>
    <arg name="robot3" default="true"/>
    <arg name="robot4" default="true"/>
    <arg name="robot5" default="false"/>

    <arg name="robot0_name" default="robot0"/>
    <arg name="robot1_name" default="robot1"/>
    <arg name="robot2_name" default="robot2"/>
    <arg name="robot3_name" default="robot3"/>
    <arg name="robot4_name" default="robot4"/>
    <arg name="robot5_name" default="robot5"/>
    
    <!-- Flags -->
    <arg name="cola2_nav" default="false"/>
    <arg name="enableNav" default="true"/>
    <arg name="enableUSBLloc" default="false"/>
    
    <!-- Interfaces -->
    <arg name="enableIqua" default="false"/>
    <arg name="enableMultimaster" default="false"/>

    <!-- ROBOT 0 -->
    <group if="$(arg robot0)">
        <include file="$(find cola2_turbot)/launch/robot.launch">
            <arg name="place" value="$(arg place)"/>
            <arg name="robot_name" value="$(arg robot0_name)"/>
            <arg name="sim" value="true"/>
            <arg name="simGPS" value="true"/>
            <arg name="enableComputerLogger" default="true"/>
            <arg name="cola2_nav" value="$(arg cola2_nav)"/>
            <arg name="enableNav" default="$(arg enableNav)"/>
            <arg name="enableSafety" value="false"/>
            <arg name="enableThrusters" value="false"/>
            <arg name="openLoopTeleoperation" value="false"/>
            <arg name="enableUSBLloc" value="$(arg enableUSBLloc)"/>
            <arg name="enableCamera" value="false"/>
            <arg name="enableModem" value="false"/>
            <arg name="enableRelayBoard" value="false"/>
            <arg name="enableIqua" value="$(arg enableIqua)"/>
            <arg name="enableMultimaster" value="$(arg enableMultimaster)"/>
            <arg name="enableEchosounder" value="false"/>
            <arg name="world2ned" value="false"/>
        </include>

        <!-- Internal Sensors -->
        <group ns="$(arg robot0_name)">
            <rosparam command="load" file="$(find cola2_turbot)/config/sim_internal_sensors.yaml"/>
            <node name="sim_internal_sensors" pkg="cola2_sim" type="sim_internal_sensors.py" output="screen"/>
            <!-- <node name="sim_pressure_sensor" pkg="cola2_turbot" type="depth_sensor_publisher.py" output="screen"/> -->
            <rosparam command="load" file="$(find cola2_turbot)/config/frames/frames_robot0.yaml"/>
            <rosparam command="load" file="$(find cola2_turbot)/config/frames/robot_localization_robot0.yaml" ns="navigator"/>
        </group>
    </group>

    <!-- ROBOT 1 -->
    <group if="$(arg robot1)">
        <include file="$(find cola2_turbot)/launch/robot.launch">
            <arg name="place" value="$(arg place)"/>
            <arg name="robot_name" value="$(arg robot1_name)"/>
            <arg name="sim" value="true"/>
            <arg name="simGPS" value="true"/>
            <arg name="enableComputerLogger" default="true"/>
            <arg name="cola2_nav" value="$(arg cola2_nav)"/>
            <arg name="enableSafety" value="false"/>
            <arg name="enableThrusters" value="false"/>
            <arg name="openLoopTeleoperation" value="false"/>
            <arg name="enableUSBLloc" value="$(arg enableUSBLloc)"/>
            <arg name="enableCamera" value="false"/>
            <arg name="enableModem" value="false"/>
            <arg name="enableRelayBoard" value="false"/>
            <arg name="enableIqua" value="$(arg enableIqua)"/>
            <arg name="enableMultimaster" value="$(arg enableMultimaster)"/>
            <arg name="enableEchosounder" value="false"/>
            <arg name="world2ned" value="false"/>
        </include>

        <!-- SIMULATE SENSORS and ACTUATORS -->
        <group ns="$(arg robot1_name)">
            <!-- Internal Sensors -->
            <rosparam command="load" file="$(find cola2_turbot)/config/sim_internal_sensors.yaml"/>
            <node name="sim_internal_sensors" pkg="cola2_sim" type="sim_internal_sensors.py" output="screen"/>
            <!-- <node name="sim_pressure_sensor" pkg="cola2_turbot" type="depth_sensor_publisher.py" output="screen"/> -->
            <rosparam command="load" file="$(find cola2_turbot)/config/frames/frames_robot1.yaml"/>
            <rosparam command="load" file="$(find cola2_turbot)/config/frames/robot_localization_robot1.yaml" ns="navigator"/>
        </group>
    </group>

    <!-- ROBOT 2 -->
    <group if="$(arg robot2)">
        <include file="$(find cola2_turbot)/launch/robot.launch">
            <arg name="place" value="$(arg place)"/>
            <arg name="robot_name" value="$(arg robot2_name)"/>
            <arg name="sim" value="true"/>
            <arg name="simGPS" value="true"/>
            <arg name="enableComputerLogger" default="true"/>
            <arg name="cola2_nav" value="$(arg cola2_nav)"/>
            <arg name="enableSafety" value="false"/>
            <arg name="enableThrusters" value="false"/>
            <arg name="openLoopTeleoperation" value="false"/>
            <arg name="enableUSBLloc" value="$(arg enableUSBLloc)"/>
            <arg name="enableCamera" value="false"/>
            <arg name="enableModem" value="false"/>
            <arg name="enableRelayBoard" value="false"/>
            <arg name="enableIqua" value="$(arg enableIqua)"/>
            <arg name="enableMultimaster" value="$(arg enableMultimaster)"/>
            <arg name="enableEchosounder" value="false"/>
            <arg name="world2ned" value="false"/>
        </include>

        <!-- SIMULATE SENSORS and ACTUATORS -->
        <group ns="$(arg robot2_name)">
            <!-- Internal Sensors -->
            <rosparam command="load" file="$(find cola2_turbot)/config/sim_internal_sensors.yaml"/>
            <node name="sim_internal_sensors" pkg="cola2_sim" type="sim_internal_sensors.py" output="screen"/>
            <!-- <node name="sim_pressure_sensor" pkg="cola2_turbot" type="depth_sensor_publisher.py" output="screen"/> -->
            <rosparam command="load" file="$(find cola2_turbot)/config/frames/frames_robot2.yaml"/>
            <rosparam command="load" file="$(find cola2_turbot)/config/frames/robot_localization_robot2.yaml" ns="navigator"/>
        </group>
    </group>

    <!-- ROBOT 3 -->
    <group if="$(arg robot3)">
        <include file="$(find cola2_turbot)/launch/robot.launch">
            <arg name="place" value="$(arg place)"/>
            <arg name="robot_name" value="$(arg robot3_name)"/>
            <arg name="sim" value="true"/>
            <arg name="simGPS" value="true"/>
            <arg name="enableComputerLogger" default="true"/>
            <arg name="cola2_nav" value="$(arg cola2_nav)"/>
            <arg name="enableSafety" value="false"/>
            <arg name="enableThrusters" value="false"/>
            <arg name="openLoopTeleoperation" value="false"/>
            <arg name="enableUSBLloc" value="$(arg enableUSBLloc)"/>
            <arg name="enableCamera" value="false"/>
            <arg name="enableModem" value="false"/>
            <arg name="enableRelayBoard" value="false"/>
            <arg name="enableIqua" value="$(arg enableIqua)"/>
            <arg name="enableMultimaster" value="$(arg enableMultimaster)"/>
            <arg name="enableEchosounder" value="false"/>
            <arg name="world2ned" value="false"/>
        </include>

        <!-- SIMULATE SENSORS and ACTUATORS -->
        <group ns="$(arg robot3_name)">
            <!-- Internal Sensors -->
            <rosparam command="load" file="$(find cola2_turbot)/config/sim_internal_sensors.yaml"/>
            <node name="sim_internal_sensors" pkg="cola2_sim" type="sim_internal_sensors.py" output="screen"/>
            <!-- <node name="sim_pressure_sensor" pkg="cola2_turbot" type="depth_sensor_publisher.py" output="screen"/> -->
            <rosparam command="load" file="$(find cola2_turbot)/config/frames/frames_robot3.yaml"/>
            <rosparam command="load" file="$(find cola2_turbot)/config/frames/robot_localization_robot3.yaml" ns="navigator"/>
        </group>
    </group>

    <!-- ROBOT 4 -->
    <group if="$(arg robot4)">
        <!-- Launch the ASV architecture -->
        <include  file="$(find multi_robot_system)/launch/ASV_simulation.launch">
            <rosparam command="load" file="$(find cola2_xiroi)/config/frames.yaml"/>
            <arg name="enableIqua" value="$(arg enableIqua)"/>
            <arg name="place" value="$(arg place)"/>
            <arg name="robot_name" value="$(arg robot4_name)"/>
        </include>
    </group>

    <!-- Run STONEFISH Simulator -->
    <arg name="simulation_data" default="$(find stonefish_ros)/simulation"/>
    <arg name="scenario_description" default="$(find stonefish_ros)/simulation/scenarios/andratx_multirobot.scn"/>
    <arg name="simulation_rate" default="500"/>
    <arg name="graphics_resolution" default="600 400"/>
    <arg name="graphics_quality" default="medium"/>
    
    <node name="stonefish_simulator" pkg="stonefish_ros" type="parsed_simulator" args="$(arg simulation_data) $(arg scenario_description) $(arg simulation_rate) $(arg graphics_resolution) $(arg graphics_quality)" output="screen">
        <remap from="stereo_down/depth/camera_info"  to="/stereo_down/depth/camera_info"/>
        <remap from="stereo_down/depth/image_depth"  to="/stereo_down/depth/image_depth"/>
        <remap from="stereo_down/left/image_color"  to="/stereo_down/left/image_raw"/>
        <remap from="stereo_down/right/image_color" to="/stereo_down/right/image_raw"/>
        <remap from="stereo_down/left/camera_info"  to="/stereo_down/left/camera_info"/>
        <remap from="stereo_down/right/camera_info" to="/stereo_down/right/camera_info0"/>
    </node>

</launch>